### DISCUSSION TITLE
Using Permalink Manager to set a custom link, the page cannot be queried through nodeByUri, how can I solve this problem?

### DISCUSSION THREAD

**Discussion by youdianfan** [wp-graphql/wp-graphql#2750] [Q&A]
I used the CPT plug-in to create an Event type, and then reset the fixed link through Permalink Manager. Then I used wp-graphql to query, but the page could not be queried. Return null

![image](https://user-images.githubusercontent.com/27018910/222968270-5e88919e-a314-4e9f-baa6-1c805acd1a66.png)

I can query the information through the old link, but it returns a new uri. I expect to query the content through `/events/test/`. Is there any way to solve this problem?

![image](https://user-images.githubusercontent.com/27018910/222968313-1983c1b3-a016-40d7-8740-1cc562eaab1a.png)

This is my Permalink Manager Settings

![image](https://user-images.githubusercontent.com/27018910/222968006-edbbb6fb-86b4-4e04-8bc9-2a6924d64314.png)

**Comment by youdianfan** [comment]
My CPT, Events Post Type Settings



``` php
function cptui_register_my_cpts_event() {
	$labels = [
		"name" => esc_html__( "Events", "twentytwentytwo" ),
		"singular_name" => esc_html__( "Event", "twentytwentytwo" ),
		"menu_name" => esc_html__( "Events", "twentytwentytwo" ),
		"all_items" => esc_html__( "All Events", "twentytwentytwo" ),
		"add_new_item" => esc_html__( "Add New Event", "twentytwentytwo" ),
		"edit_item" => esc_html__( "Edit Event", "twentytwentytwo" ),
		"new_item" => esc_html__( "New Event", "twentytwentytwo" ),
		"view_item" => esc_html__( "View Event", "twentytwentytwo" ),
		"view_items" => esc_html__( "View Events", "twentytwentytwo" ),
		"search_items" => esc_html__( "Search Event", "twentytwentytwo" ),
		"not_found" => esc_html__( "No Events found", "twentytwentytwo" ),
		"not_found_in_trash" => esc_html__( "No Events found in Trash", "twentytwentytwo" ),
		"parent" => esc_html__( "Parent Event:", "twentytwentytwo" ),
		"featured_image" => esc_html__( "Featured image for this event", "twentytwentytwo" ),
		"parent_item_colon" => esc_html__( "Parent Event:", "twentytwentytwo" ),
	];

	$args = [
		"label" => esc_html__( "Events", "twentytwentytwo" ),
		"labels" => $labels,
		"description" => "",
		"public" => true,
		"publicly_queryable" => true,
		"show_ui" => true,
		"show_in_rest" => true,
		"rest_base" => "",
		"rest_controller_class" => "WP_REST_Posts_Controller",
		"rest_namespace" => "wp/v2",
		"has_archive" => false,
		"show_in_menu" => true,
		"show_in_nav_menus" => true,
		"delete_with_user" => false,
		"exclude_from_search" => false,
		"capability_type" => "post",
		"map_meta_cap" => true,
		"hierarchical" => false,
		"can_export" => true,
		"rewrite" => [ "slug" => "event", "with_front" => true ],
		"query_var" => true,
		"menu_icon" => "dashicons-calendar-alt",
		"supports" => [ "title", "editor", "thumbnail", "excerpt", "custom-fields" ],
		"show_in_graphql" => true,
		"graphql_single_name" => "Event",
		"graphql_plural_name" => "Events",
	];

	register_post_type( "event", $args );
}

add_action( 'init', 'cptui_register_my_cpts_event' );
```

**Comment by youdianfan** [comment]
My Permalink Manager Settings
`events/%event%`
![image](https://user-images.githubusercontent.com/27018910/222968625-b4d633ca-f5d4-43a0-8bc2-bf66eb0012bc.png)

**Comment by youdianfan** [comment]
My plugin information

![image](https://user-images.githubusercontent.com/27018910/222968781-81454101-7813-405a-9edb-0cad469fcfbe.png)

**Comment by justlevine** [comment]
Hey @youdianfan , Permalink Manager is not supported by WPGraphQL. It's possible that the upcoming release ( #2748 ) will add compatibility, but I wouldnt bet on it, since I'm pretty sure Permalink Manager does some non-standard things to manage the redirects.

**Reply by slowdub** [reply]
Confirming that it won't work (at least not with the current pre-release version I tested with).
If permalink manager plugin doesn't work, what's the recommend way to achieve custom permalinks on a per page/post basis which is compatible with WPGraphQL? Or at least on a CPT basis (without forcing to use CPT slug prefix)?

**Reply by justlevine** [reply]
@blistok filtering the WP rewrite rules (or a plugin that does it for you) for the per-page rules. If you just want to strip a CPT prefix, you can set `$args['rewrite']['with_front']` to false.

Ref: https://developer.wordpress.org/reference/functions/register_post_type/#rewrite

**Reply by slowdub** [reply]
@justlevine I don't think this is doing what we're discussing here. Setting `$args['rewrite']['with_front']` to `false` only removes the static prefix (in the case it's set in Settings -> Permalinks -> Custom Structure), but it doesn't remove the CPT "slug" prefix. 

If I have a CPT Portfolio, with enabled rewrite and slug `portfolio`, the URL structure would look like:
`/portfolio/%postname%`
If let's say structure in the Settings -> Permalinks is set to: `/my-static-prefix/%postname%`
the URL structure for Portfolio CPT will be:
`/my-static-prefix/portfolio/%postname%` if `$args['rewrite']['with_front']` is `true` and:
`/portfolio/%postname%` if it's `false`.

So, the `/portfolio/` prefix stays either way.

Am I wrong here?

**Reply by justlevine** [reply]
No you're not. I misunderstood what you meant by the CPT slug prefix. The best way to handle those is to add custom rules to WP_Rewrite or use one of the NodeResolver class filters.
